sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala

```diff
@@ -324,20 +324,24 @@ class CertificateSubmitter(settings: SidechainSettings,
     val privateKeysWithIndexes = Await.result(sidechainNodeViewHolderRef ? GetDataFromCurrentView(getSignersPrivateKeys), timeoutDuration)
       .asInstanceOf[Seq[(SchnorrSecret, Int)]]
 
-    (signaturesFromEnclave(messageToSign, params.signersPublicKeys.zipWithIndex)
+    val locallyPresentKeysIndexes = privateKeysWithIndexes.map(_._2)
+
+    (signaturesFromEnclave(messageToSign, params.signersPublicKeys.zipWithIndex.filterNot(key_index => locallyPresentKeysIndexes.contains(key_index._2)))
       ++ privateKeysWithIndexes.map {
       case (secret, pubKeyIndex) => CertificateSignatureInfo(pubKeyIndex, secret.sign(messageToSign))
     })
-      .distinct
   }
 
   def signaturesFromEnclave(messageToSign: Array[Byte], indexedPublicKeys: Seq[(SchnorrProposition, Int)]): Seq[CertificateSignatureInfo] = {
-    Await.result(
-      Future.sequence(
-        indexedPublicKeys.map(publicKey_index => secureEnclaveApiClient ? SignWithEnclave(messageToSign, publicKey_index))
-      ), timeoutDuration
-    ).asInstanceOf[Seq[Option[CertificateSignatureInfo]]]
+    if (!secureEnclaveApiClient.isEnabled) return Seq()
+
+    val signaturesFromEnclaveFuture = secureEnclaveApiClient.listPublicKeys()
+      .map(managedKeys => indexedPublicKeys.filter(key_index => managedKeys.contains(key_index._1)))
+      .map(_.map(secureEnclaveApiClient.signWithEnclave(messageToSign, _)))
+      .map(Future.sequence(_))
       .flatten
+
+    Await.result(signaturesFromEnclaveFuture, timeoutDuration).flatten
```

Comment:
does it cover Timeout exception case?
So in case we have timeout, we return empty sequence rather than passing exception further.

Reply:
In case of a timeout, we'll hit this part inside API client
```
        case response: HttpResponse =>
          logger.error(s"Secure Enclave returned non-200 response: ${response.status}, ${response.entity}")
          Future(Seq())
```
and immediately return empty seq here

akka timeout can be configured using `akka.http.server.request-timeout` config setting, and is 20 sec by default

One thing I missed is that one can set restApi timeout to be something smaller, so I'll cover it to work as you said